packages/@lwc/template-compiler/src/shared/ir.ts

```diff
@@ -106,5 +105,8 @@ export function isBoundToIterator(identifier: TemplateIdentifier, node?: IRNode)
     }
 
     // Delegate to parent component if no binding is found at this point
-    return isBoundToIterator(identifier, node.parent);
+    return isBoundToIterator(identifier, irParentMap.get(node));
 }
+
+export const irParentMap: WeakMap<IRElement | IRText, IRElement> = new WeakMap();
```

Comment:
I am really not a fan of having globals in the compiler, the AST and the parser internal state should keep track of the mapping.